core/src/task/tick.rs

```diff
@@ -0,0 +1,58 @@
+use std::{sync::Arc, time::Duration};
+use tokio::select;
+use triggered::{trigger, Listener, Trigger};
+
+use super::service::{AsyncService, AsyncServiceFuture};
+
+const TICK: &str = "tick";
+
+pub struct TickService {
+    shutdown_trigger: Trigger,
+    shutdown_listener: Listener,
+}
+
+impl TickService {
+    pub fn new() -> Self {
+        let (shutdown, monitor) = trigger();
+        Self { shutdown_trigger: shutdown, shutdown_listener: monitor }
+    }
+
+    /// Waits until `duration` has elapsed when the service is started.
+    ///
+    /// Returns immediately when the service is stopped.
+    pub async fn tick(&self, duration: Duration) {
+        let shutdown_listener = self.shutdown_listener.clone();
+        loop {
```

Comment:
why the loop?

Reply:
btw you can also await the listener via `tokio::time::timeout` instead of using a `select`